src/java.base/share/classes/java/lang/StackStreamFactory.java

```diff
@@ -415,31 +423,36 @@ private R beginStackWalk() {
 
             return callStackWalk(mode, 0,
                                  contScope, continuation,
-                                 frameBuffer.curBatchFrameCount(),
+                                 frameBuffer.currentBatchSize(),
                                  frameBuffer.startIndex(),
                                  frameBuffer.frames());
         }
 
         /*
-         * Fetches stack frames.
+         * Fetches a new batch of stack frames.  This method returns
+         * the number of stack frames filled in this batch.
          *
-         * @param batchSize number of elements of the frame buffers for this batch
-         * @return number of frames fetched in this batch
+         * When it reaches the bottom of a continuation, i.e. Continuation::enter,
+         * VM ends the batch and let the stack walker to set the continuation
+         * to its parent and continue the stack walking.  It may return zero.
          */
-        private int fetchStackFrames(int batchSize) {
+        private int fetchStackFrames() {
             int startIndex = frameBuffer.startIndex();
-            frameBuffer.resize(startIndex, batchSize);
 
-            int endIndex = fetchStackFrames(mode, anchor, batchSize,
-                                            startIndex,
+            // If the last batch didn't fetch any frames, keep the current batch size.
+            int lastBatchFrameCount = frameBuffer.numFrames();
```

Comment:
I run some tests to understand the issue and I got the same failure if I now set MIN_BATCH_SIZE to 7. This just forces the same situation where Continuation::enter is the last frame in the buffer, otherwise since the patch also changes the batch sizes we don't fall into that issue anymore. The problem is with this numFrames() method which still returns a number > 0 after the fetch attempt that returns with no frames. Maybe there is a reset missing for origin and fence when fetching the next batch?

Reply:
Thanks for catching this.   The problem is that `fetchStackFrames(int batchSize)` is supposed to call `setBatch` to set origin and fence for the new batch.  But if no frame is fetched, it skips not calling that.   I have a fix for it.